### ISSUE TITLE
More intuitive support for referencing orgs/lbs etc

### ISSUE THREAD

**Issue by strideynet** [krystal/go-katapult#96]
At the moment, when interacting with or referring to resources we tend to require the consumer to pass the entire object in that they want to refer to. This is pretty inconvenient as it will force the user to either

1. Work out which fields are actually being used, and fill those from known values (e.g ID or subdomain)
2. Fetch the entire object before interacting with/referencing it

I'd like to suggest probably a cleaner interface along the lines of:

```golang
type OrgReferencer func(qs *url.Values)

func ByOrgID(orgId string) OrgReferencer {
	return func(qs *url.Values) {
		qs.Set("organization[id]", orgId)
	}
}
func ByOrgSubdomain(subdomain string) OrgReferencer {
	return func(qs *url.Values) {
		qs.Set("organization[sub_domain]", subdomain)
	}
}

func (o Organization) Ref() OrgReferencer {
	return func(qs *url.Values) {
		qs.Set("organization[id]", o.ID)
	}
}

// Lists the VMs for an org
func (s *VirtualMachinesClient) List(
	ctx context.Context,
	orgRef OrgReferencer,
	opts *ListOptions,
) ([]*VirtualMachine, *katapult.Response, error) {
	qs := queryValues(opts)
	orgRef(qs)
	u := &url.URL{
		Path:     "organizations/_/virtual_machines",
		RawQuery: qs.Encode(),
	}
	body, resp, err := s.doRequest(ctx, "GET", u, nil)
	resp.Pagination = body.Pagination
	return body.VirtualMachines, resp, err
}
```

This gives a consumer experience along the lines of:

```
c.LoadBalancers.List(ctx, ByOrgId(orgID})
c.LoadBalancers.List(ctx, ByOrgSubdomain(orgSubDomain))
org := &Organization{}
c.LoadBalancers.List(ctx, org.Ref())
```

I'm more certain of how I want the consumer experience to be than the actually underlying implementation, which I feel could equally be implemented using aliased function type (as shown in the example) or by returning a type that implements the queryable interface.

**Comment by jimeh** [comment]
Your example makes sense, and gives a solid type to the currently vague and annoying concept of a "lookup reference" of all the API resource types.

I would like to tweaks names a bit though, calling the ref type `OrganizationRef`, basically including the full name of the type that its a reference for. And it'll make `organization.Ref()` feel more natural too, as it returns a `OrganizationRef`.

And similarly with the root level functions, I'd prefer `OrganizationRefByID()`, `OrganizationRefBySubDomain()`, and maybe even prefix both with `New` depending on how godoc is capable of grouping things together.

As for the underlying implementation, your func-based example here is neat, but would limit us exclusively to query strings. A more flexible solution might be to simply make it a struct with only the fields available to lookup the resource by, which might actually help making the docs even clearer about what fields can be used.

**Comment by strideynet** [comment]
I like the suggestion to have some form of a struct returned that implements the Queryable interface. As you say, it restricts us less in future if we do decide to change something here.

**Comment by jimeh** [comment]
Yeah, I think we should still have `OrganizationRef` and all `*Ref` structs implement the `queryable` interface, as that's the primary way we pass these values to the API. 

But I just realized there's some places where we definitely need the refs marshaled to JSON too, a prime example is when [creating a Load Balancer](https://developers.katapult.io/docs/api/latest/core/routes/organizations/:organization/load_balancers/post/), the `properties` post body attribute contains a `data_center` attribute which is a [`DataCenterLookup`](https://developers.katapult.io/docs/api/latest/core/argument_sets/a8e39cf1be/) type, for which we'd not want to send a full `DataCenter` JSON blob.
